### PR TITLE
GH-98831: Refactor and fix cases generator

### PR DESCRIPTION
**Pull request by gvanrossum**
This refactor also fixes code generation for cache effects that aren't unused and removes `check_overlaps()`. It shows it all off for the `BINARY_SUBSCR` family. (I had an earlier PR (GH-99408) for this, but it got a bit stale and I decided to delete the history.)

* Issue: gh-98831

### CODE REVIEW

**Comment by gvanrossum** [comment]
I have a newer version of the same code in GH-99495 (which adds macros and ops) but that's not fully baked and I figured I'd release this in stages so it's easier on the reviewer. If you'd rather review the tooling in even larger chunks let me know. (I think the changes to the instruction definitions are pretty mild in this one.)

File: Python/bytecodes.c [hunk 1]
```diff
@@ -410,51 +409,39 @@ dummy_func(
...
+            Py_DECREF(sub);
         }
 
-        // stack effect: (__0 -- )
-        inst(BINARY_SUBSCR_GETITEM) {
-            PyObject *sub = TOP();
-            PyObject *container = SECOND();
-            _PyBinarySubscrCache *cache = (_PyBinarySubscrCache *)next_instr;
-            uint32_t type_version = read_u32(cache->type_version);
+        inst(BINARY_SUBSCR_GETITEM, (container, sub, unused/1, type_version/2, func_version/1 -- unused)) {
```
Comment (brandtbucher) [inline Python/bytecodes.c:436]: Super cool to see the cache reads in action!

File: Python/bytecodes.c [hunk 2]
```diff
@@ -3197,7 +3184,7 @@ dummy_func(
                 goto error;
             }
             PyObject *arg = TOP();
-            PyObject *res = _PyCFunction_TrampolineCall(cfunc, PyCFunction_GET_SELF(callable), arg);
+            PyObject *res = cfunc(PyCFunction_GET_SELF(callable), arg);
```
Comment (brandtbucher) [inline Python/bytecodes.c]: Did you mean to change this (also the two instances below)?

File: Tools/cases_generator/parser.py [hunk 3]
```diff
@@ -90,13 +89,12 @@ def name(self) -> str:
         return self.header.name
 
     @property
-    def inputs(self) -> list[Effect]:
+    def inputs(self) -> list[InputEffect]:
         return self.header.inputs
 
     @property
     def outputs(self) -> list[StackEffect]:
```
Comment (brandtbucher) [inline Tools/cases_generator/parser.py:96]:
```suggestion
    def outputs(self) -> list[OutputEffect]:
```

File: Tools/cases_generator/generate_cases.py [hunk 4]
```diff
@@ -1,55 +1,326 @@
...
+                        f"{indent}    static_assert({cache_size} == "
+                        f'{self.cache_offset}, "incorrect cache size");\n'
+                    )
+
+        # Write cache effect variable declarations
+        cache_offset = 0
+        for ceffect in self.cache_effects:
+            if ceffect.name != "unused":
+                # TODO: if name is 'descr' use PyObject *descr = read_obj(...)
+                bits = ceffect.size * 16
```
Comment (brandtbucher) [inline Tools/cases_generator/generate_cases.py:75]: I had suggestions on these from the other PR, not sure if you saw (I like `_` more than `unused`, and both `"unused"` and `16` might benefit from constant names in this file).

Unless you just disagreed, in which case, feel free to ignore this. :)

File: Tools/cases_generator/generate_cases.py [hunk 5]
```diff
@@ -1,55 +1,326 @@
...
+        diff = len(self.output_effects) - len(self.input_effects)
+        if diff > 0:
+            f.write(f"{indent}    STACK_GROW({diff});\n")
+        elif diff < 0:
+            f.write(f"{indent}    STACK_SHRINK({-diff});\n")
+
+        # Write output stack effect assignments
+        input_names = [seffect.name for seffect in self.input_effects]
+        for i, output in enumerate(reversed(self.output_effects), 1):
+            if output.name not in input_names and output.name != "unused":
```
Comment (brandtbucher) [inline Tools/cases_generator/generate_cases.py:110]: For `output.name not in input_names`, don't we need to check that they occur at the same (stack-adjusted) position, and write them out if not?

File: Tools/cases_generator/generate_cases.py [hunk 6]
```diff
@@ -1,55 +1,326 @@
...
+        psr = parser.Parser(self.src, filename=self.filename)
+
+        # Skip until begin marker
+        while tkn := psr.next(raw=True):
+            if tkn.text == BEGIN_MARKER:
+                break
         else:
-            raise psr.make_syntax_error(f"Unexpected token")
-    return instrs, supers, families
+            raise psr.make_syntax_error(f"Couldn't find {BEGIN_MARKER!r} in {psr.filename}")
```
Comment (brandtbucher) [inline Tools/cases_generator/generate_cases.py:183]: Wild, I've never seen `while`/`else` before...

File: Tools/cases_generator/generate_cases.py [hunk 7]
```diff
@@ -1,55 +1,326 @@
...
+        for instr in self.instrs.values():
+            for target in re.findall(RE_PREDICTED, instr.block.text):
+                if target_instr := self.instrs.get(target):
+                    target_instr.predicted = True
+                else:
+                    print(
+                        f"Unknown instruction {target!r} predicted in {instr.name!r}",
+                        file=sys.stderr,
+                    )
+                    self.errors += 1
```
Comment (brandtbucher) [inline Tools/cases_generator/generate_cases.py:226]: You do this move in lots of places... maybe add a method?

```py
def error(self, e: str) -> None:
    print(e, file=sys.stderr)
    self.errors += 1
```

File: Tools/cases_generator/generate_cases.py [hunk 8]
```diff
@@ -1,55 +1,326 @@
...
+            if len(family.members) < 2:
+                print(f"Family {family.name!r} has insufficient members")
+                self.errors += 1
+            members = [member for member in family.members if member in self.instrs]
+            if members != family.members:
+                unknown = set(family.members) - set(members)
+                print(f"Family {family.name!r} has unknown members: {unknown}")
+                self.errors += 1
+            if len(members) < 2:
+                continue
```
Comment (brandtbucher) [inline Tools/cases_generator/generate_cases.py:258]: This should be an error, yeah?

**Review by brandtbucher — APPROVED** [review]
Looks good. A few comments:

**Pull request merged**